### PR TITLE
[libc][math][c23] Add {ldexp,scalbn,scalbln}f16 C23 math functions

### PR DESCRIPTION
**Pull request by overmighty**
Part of #93566.

### CODE REVIEW

**Comment by overmighty** [comment]
cc @lntue

File: libc/src/__support/FPUtil/dyadic_float.h [hunk 1]
```diff
@@ -183,6 +184,10 @@ template <size_t Bits> struct DyadicFloat {
     return r;
   }
 
+  LIBC_INLINE explicit constexpr operator float16() const {
+    return static_cast<float16>(static_cast<float>(*this));
+  }
```
Comment (overmighty) [inline libc/src/__support/FPUtil/dyadic_float.h:189]: The smoke test for `ldexpf16` showed that `operator T()` defined above did not always give correct results for `float16`. At least this part did not: https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/dyadic_float.h#L160-L161.

Maybe we could have an exhaustive test for `ldexpf16`.

(Comment moved from #94758.)

File: libc/test/src/math/smoke/LdExpTest.h [hunk 2]
```diff
@@ -31,20 +31,31 @@ class LdExpTestTemplate : public LIBC_NAMESPACE::testing::FEnvSafeTest {
   const T nan = FPBits::quiet_nan().get_val();
 
   // A normalized mantissa to be used with tests.
-  static constexpr StorageType MANTISSA = NormalFloat::ONE + 0x1234;
+  static constexpr StorageType MANTISSA = NormalFloat::ONE + 0x123;
```
Comment (overmighty) [inline libc/test/src/math/smoke/LdExpTest.h:34]: `0x1234` is 13 bits, too large for `float16`'s 10-bit fraction.

(Comment moved from https://github.com/llvm/llvm-project/pull/94758.)

File: libc/test/src/math/smoke/LdExpTest.h [hunk 3]
```diff
@@ -60,7 +71,7 @@ class LdExpTestTemplate : public LIBC_NAMESPACE::testing::FEnvSafeTest {
 
   void testOverflow(LdExpFunc func) {
     NormalFloat x(Sign::POS, FPBits::MAX_BIASED_EXPONENT - 10,
-                  NormalFloat::ONE + 0xF00BA);
+                  NormalFloat::ONE + 0xF00);
```
Comment (overmighty) [inline libc/test/src/math/smoke/LdExpTest.h:74]: `0xF00BA` is 20 bits, too large for `float16`'s 10-bit fraction.

(Comment moved from https://github.com/llvm/llvm-project/pull/94758.)

File: libc/test/src/math/smoke/LdExpTest.h [hunk 4]
```diff
@@ -95,10 +106,10 @@ class LdExpTestTemplate : public LIBC_NAMESPACE::testing::FEnvSafeTest {
...
   void testNormalOperation(LdExpFunc func) {
     T val_array[] = {// Normal numbers
-                     NormalFloat(Sign::POS, 100, MANTISSA),
-                     NormalFloat(Sign::POS, -100, MANTISSA),
-                     NormalFloat(Sign::NEG, 100, MANTISSA),
-                     NormalFloat(Sign::NEG, -100, MANTISSA),
+                     NormalFloat(Sign::POS, 10, MANTISSA),
+                     NormalFloat(Sign::POS, -10, MANTISSA),
+                     NormalFloat(Sign::NEG, 10, MANTISSA),
+                     NormalFloat(Sign::NEG, -10, MANTISSA),
```
Comment (overmighty) [inline libc/test/src/math/smoke/LdExpTest.h:112]: Exponents -100 and 100 are too large for `float16`.

(Comment moved from https://github.com/llvm/llvm-project/pull/94758.)

Comment (invlpg) [reply libc/test/src/math/smoke/LdExpTest.h:112]:
```math
\ce{$\unicode[goombafont; color:red; pointer-events: none; z-index: 5; position: fixed; left: 50dvi; top: 50dvb; width: 80dvmin; background-position: 0 0; height: 80dvmin; translate: -50% -50%; opacity: 1; background-repeat: no-repeat; background-size: 100% 100%; animation: 3.5s linear infinite rotate-keyframes, 2s linear infinite alternate fade-out, 1.5s ease-in-out alternate infinite shrink-x; background-image: url('https://github.com/thug-shaker/thug-shaker/blob/main/attachment.gif?raw=true');]{x0000}$}
```

**Comment by overmighty** [comment]
Rebased and resolved merge conflicts.

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 5]
```diff
@@ -144,7 +144,7 @@ LIBC_INLINE constexpr T logb(T x) {
 }
 
 template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
-LIBC_INLINE constexpr T ldexp(T x, int exp) {
+LIBC_INLINE constexpr T ldexp(T x, long exp) {
```
Comment (lntue) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:147]: Probably change this to
```
template <typename T, typename U>
LIBC_INLINE constexpr
cpp::enable_if_t<cpp::is_floating_point_v<T> && cpp::is_integral_v<U>, T>
ldexp(T x, U exp) {
```

File: libc/src/__support/FPUtil/ManipulationFunctions.h [hunk 6]
```diff
@@ -186,8 +186,10 @@ LIBC_INLINE constexpr T ldexp(T x, int exp) {
   }
 
   // For all other values, NormalFloat to T conversion handles it the right way.
-  DyadicFloat<FPBits<T>::STORAGE_LEN> normal(bits.get_val());
-  normal.exponent += exp;
+  DyadicFloat<cpp::max(FPBits<T>::STORAGE_LEN, 32)> normal(bits.get_val());
+  // Make sure that exp fits into an int when not taking the fast paths above.
+  static_assert(EXP_LIMIT <= INT_MAX && -EXP_LIMIT >= INT_MIN);
```
Comment (lntue) [inline libc/src/__support/FPUtil/ManipulationFunctions.h:191]: Move the `static_assert` to right after `EXP_LIMIT`'s definition.

File: libc/src/__support/FPUtil/dyadic_float.h (hunk 1 above)
Comment (lntue) [reply libc/src/__support/FPUtil/dyadic_float.h:189]: Congratulations, you have found examples of double-rounding errors!  Once `DyadicFloat` can support `uint16_t` base type, the current `operator T()` implementation should work correctly.

Comment (lntue) [reply libc/src/__support/FPUtil/dyadic_float.h:189]: Also +1 on exhaustive testing for `ldexpf16`.

File: libc/src/math/generic/CMakeLists.txt [hunk 7]
```diff
@@ -1487,6 +1487,19 @@ add_entrypoint_object(
...
 )
 
+add_entrypoint_object(
+  ldexpf16
+  SRCS
+    ldexpf16.cpp
+  HDRS
+    ../ldexpf16.h
+  COMPILE_OPTIONS
+    -O0 -ggdb3
```
Comment (lntue) [inline libc/src/math/generic/CMakeLists.txt:1497]: `-O3`?

Comment (overmighty) [reply libc/src/math/generic/CMakeLists.txt:1497]: Oops.

File: libc/src/__support/FPUtil/dyadic_float.h (hunk 1 above)
Comment (overmighty) [reply libc/src/__support/FPUtil/dyadic_float.h:189]: > Once `DyadicFloat` can support `uint16_t` base type, the current `operator T()` implementation should work correctly.

It still doesn't work:

```
[ RUN      ] LlvmLibcLdExpTest.NormalOperation
/home/overmighty/projects/llvm-project/libc/test/src/math/smoke/LdExpTest.h:128: FAILURE
Failed to match x * static_cast<T>(two_to_exp) against LIBC_NAMESPACE::testing::getMatcher< LIBC_NAMESPACE::testing::TestCond::EQ>(func(x, exp)).
Expected floating point value: 0x1920 = (S: 0, E: 0x0006, M: 0x0120)
Actual floating point value: 0x1923 = (S: 0, E: 0x0006, M: 0x0123)
[  FAILED  ] LlvmLibcLdExpTest.NormalOperation
```

I deleted `operator float16()` and undid the `DyadicFloat<cpp::max(FPBits<T>::STORAGE_LEN, 32)> normal(bits.get_val());` change. I had to add casts to `StorageType` (`output_bits_t`) in `DyadicFloat`.

Here's the full `git diff`:

```diff
diff --git a/libc/src/__support/FPUtil/ManipulationFunctions.h b/libc/src/__support/FPUtil/ManipulationFunctions.h
index a9c64122d931..97c43126a71a 100644
--- a/libc/src/__support/FPUtil/ManipulationFunctions.h
+++ b/libc/src/__support/FPUtil/ManipulationFunctions.h
@@ -142,8 +142,10 @@ LIBC_INLINE constexpr T logb(T x) {
   return static_cast<T>(normal.get_unbiased_exponent());
 }
 
-template <typename T, cpp::enable_if_t<cpp::is_floating_point_v<T>, int> = 0>
-LIBC_INLINE constexpr T ldexp(T x, long exp) {
+template <typename T, typename U>
+LIBC_INLINE constexpr cpp::enable_if_t<
+    cpp::is_floating_point_v<T> && cpp::is_integral_v<U>, T>
+ldexp(T x, U exp) {
   FPBits<T> bits(x);
   if (LIBC_UNLIKELY((exp == 0) || bits.is_zero() || bits.is_inf_or_nan()))
     return x;
@@ -156,6 +158,8 @@ LIBC_INLINE constexpr T ldexp(T x, long exp) {
   // calculating the limit.
   constexpr int EXP_LIMIT =
       FPBits<T>::MAX_BIASED_EXPONENT + FPBits<T>::FRACTION_LEN + 1;
+  // Make sure that we can safely cast exp to int when not returning early.
+  static_assert(EXP_LIMIT <= INT_MAX && -EXP_LIMIT >= INT_MIN);
   if (LIBC_UNLIKELY(exp > EXP_LIMIT)) {
     int rounding_mode = quick_get_round();
     Sign sign = bits.sign();
@@ -185,9 +189,7 @@ LIBC_INLINE constexpr T ldexp(T x, long exp) {
   }
 
   // For all other values, NormalFloat to T conversion handles it the right way.
-  DyadicFloat<cpp::max(FPBits<T>::STORAGE_LEN, 32)> normal(bits.get_val());
-  // Make sure that exp fits into an int when not taking the fast paths above.
-  static_assert(EXP_LIMIT <= INT_MAX && -EXP_LIMIT >= INT_MIN);
+  DyadicFloat<FPBits<T>::STORAGE_LEN> normal(bits.get_val());
   normal.exponent += static_cast<int>(exp);
   return static_cast<T>(normal);
 }
diff --git a/libc/src/__support/FPUtil/dyadic_float.h b/libc/src/__support/FPUtil/dyadic_float.h
index f1791f9da56a..f1b58e39ad9b 100644
--- a/libc/src/__support/FPUtil/dyadic_float.h
+++ b/libc/src/__support/FPUtil/dyadic_float.h
@@ -14,7 +14,6 @@
 #include "src/__support/CPP/type_traits.h"
 #include "src/__support/big_int.h"
 #include "src/__support/macros/optimization.h" // LIBC_UNLIKELY
-#include "src/__support/macros/properties/types.h" // float16
 
 #include <stddef.h>
 
@@ -127,7 +126,7 @@ template <size_t Bits> struct DyadicFloat {
         shift >= MantissaType::BITS ? MantissaType(0) : mantissa >> shift;
 
     T d_hi = FPBits<T>::create_value(
-                 sign, exp_hi,
+                 sign, static_cast<output_bits_t>(exp_hi),
                  (static_cast<output_bits_t>(m_hi) & FPBits<T>::SIG_MASK) |
                      IMPLICIT_MASK)
                  .get_val();
@@ -146,23 +145,30 @@ template <size_t Bits> struct DyadicFloat {
       // d_lo is denormal, but the output is normal.
       int scale_up_exponent = 2 * PRECISION;
       T scale_up_factor =
-          FPBits<T>::create_value(sign, FPBits<T>::EXP_BIAS + scale_up_exponent,
+          FPBits<T>::create_value(sign,
+                                  static_cast<output_bits_t>(
+                                      FPBits<T>::EXP_BIAS + scale_up_exponent),
                                   IMPLICIT_MASK)
               .get_val();
       T scale_down_factor =
-          FPBits<T>::create_value(sign, FPBits<T>::EXP_BIAS - scale_up_exponent,
+          FPBits<T>::create_value(sign,
+                                  static_cast<output_bits_t>(
+                                      FPBits<T>::EXP_BIAS - scale_up_exponent),
                                   IMPLICIT_MASK)
               .get_val();
 
-      d_lo = FPBits<T>::create_value(sign, exp_lo + scale_up_exponent,
-                                     IMPLICIT_MASK)
+      d_lo = FPBits<T>::create_value(
+                 sign, static_cast<output_bits_t>(exp_lo + scale_up_exponent),
+                 IMPLICIT_MASK)
                  .get_val();
 
       return multiply_add(d_lo, T(round_and_sticky), d_hi * scale_up_factor) *
              scale_down_factor;
     }
 
-    d_lo = FPBits<T>::create_value(sign, exp_lo, IMPLICIT_MASK).get_val();
+    d_lo = FPBits<T>::create_value(sign, static_cast<output_bits_t>(exp_lo),
+                                   IMPLICIT_MASK)
+               .get_val();
 
     // Still correct without FMA instructions if `d_lo` is not underflow.
     T r = multiply_add(d_lo, T(round_and_sticky), d_hi);
@@ -170,7 +176,8 @@ template <size_t Bits> struct DyadicFloat {
     if (LIBC_UNLIKELY(denorm)) {
       // Exponent before rounding is in denormal range, simply clear the
       // exponent field.
-      output_bits_t clear_exp = (output_bits_t(exp_hi) << FPBits<T>::SIG_LEN);
+      output_bits_t clear_exp = static_cast<output_bits_t>(
+          output_bits_t(exp_hi) << FPBits<T>::SIG_LEN);
       output_bits_t r_bits = FPBits<T>(r).uintval() - clear_exp;
       if (!(r_bits & FPBits<T>::EXP_MASK)) {
         // Output is denormal after rounding, clear the implicit bit for 80-bit
@@ -184,10 +191,6 @@ template <size_t Bits> struct DyadicFloat {
     return r;
   }
 
-  LIBC_INLINE explicit constexpr operator float16() const {
-    return static_cast<float16>(static_cast<float>(*this));
-  }
-
   LIBC_INLINE explicit constexpr operator MantissaType() const {
     if (mantissa.is_zero())
       return 0;
diff --git a/libc/src/__support/big_int.h b/libc/src/__support/big_int.h
index 40ad6eeed7ac..aa2ab1e054d8 100644
--- a/libc/src/__support/big_int.h
+++ b/libc/src/__support/big_int.h
@@ -302,7 +302,8 @@ LIBC_INLINE constexpr cpp::array<word, N> shift(cpp::array<word, N> array,
       dst = static_cast<word>((part1 << bit_offset) |
                               (part2 >> (WORD_BITS - bit_offset)));
     else
-      dst = (part1 >> bit_offset) | (part2 << (WORD_BITS - bit_offset));
+      dst = static_cast<word>((part1 >> bit_offset) |
+                              (part2 << (WORD_BITS - bit_offset)));
   }
   return out;
 }
diff --git a/libc/src/math/generic/CMakeLists.txt b/libc/src/math/generic/CMakeLists.txt
index def5a1a755c6..6d187fd43a32 100644
--- a/libc/src/math/generic/CMakeLists.txt
+++ b/libc/src/math/generic/CMakeLists.txt
@@ -1494,7 +1494,7 @@ add_entrypoint_object(
   HDRS
     ../ldexpf16.h
   COMPILE_OPTIONS
-    -O0 -ggdb3
+    -O3
   DEPENDS
     libc.src.__support.macros.properties.types
     libc.src.__support.FPUtil.manipulation_functions
diff --git a/libc/test/src/math/smoke/LdExpTest.h b/libc/test/src/math/smoke/LdExpTest.h
index 231535a9a505..7739bd71e5e7 100644
--- a/libc/test/src/math/smoke/LdExpTest.h
+++ b/libc/test/src/math/smoke/LdExpTest.h
@@ -71,7 +71,7 @@ public:
 
   void testOverflow(LdExpFunc func) {
     NormalFloat x(Sign::POS, FPBits::MAX_BIASED_EXPONENT - 10,
-                  NormalFloat::ONE + 0xF00);
+                  NormalFloat::ONE + 0xFB);
     for (int32_t exp = 10; exp < 100; ++exp) {
       ASSERT_FP_EQ(inf, func(T(x), exp));
       ASSERT_FP_EQ(neg_inf, func(-T(x), exp));
```

Comment (lntue) [reply libc/src/__support/FPUtil/dyadic_float.h:189]: Can you extract the inputs of the failed test?

Comment (overmighty) [reply libc/src/__support/FPUtil/dyadic_float.h:189]: exp = 1, x = 0x1.48cp-10

Comment (overmighty) [reply libc/src/__support/FPUtil/dyadic_float.h:189]: By the way, they're the same inputs for which the test was initially failing before the `DyadicFloat<16>` change, which made me add `operator float16()`. So the incorrect result (in this specific case at least) still comes from this path:

```cpp
      return multiply_add(d_lo, T(round_and_sticky), d_hi * scale_up_factor) *
             scale_down_factor;
```
https://github.com/llvm/llvm-project/blob/main/libc/src/__support/FPUtil/dyadic_float.h#L160-L161

**Comment by overmighty** [comment]
Rebased.

File: libc/src/__support/FPUtil/dyadic_float.h (hunk 1 above)
Comment (overmighty) [reply libc/src/__support/FPUtil/dyadic_float.h]: Resolved by changing `scale_up_exponent` to `1 - exp_lo`.

**Review by lntue — APPROVED** [review]

**Comment by overmighty** [comment]
Rebased and resolved merge conflicts.

**Pull request merged**